### DISCUSSION TITLE
Problem with compiling.

### DISCUSSION THREAD

**Discussion by prakhar114** [eclipse-ecal/ecal#881] [General]
I'm getting following error while doing a cmake in protobuf_snd. I've copied and code and trying to cmake it but it's throwing me following

```
CMake Error at /usr/lib/x86_64-linux-gnu/cmake/eCAL/eCALConfig.cmake:69 (include):
  include could not find load file:

    /usr/local/lib/x86_64-linux-gnu/cmake/eCAL/helper_functions/ecal_add_functions.cmake
Call Stack (most recent call first):
  CMakeLists.txt:9 (find_package)


CMake Error at /usr/lib/x86_64-linux-gnu/cmake/eCAL/eCALConfig.cmake:70 (include):
  include could not find load file:

    /usr/local/lib/x86_64-linux-gnu/cmake/eCAL/helper_functions/ecal_helper_functions.cmake
Call Stack (most recent call first):
  CMakeLists.txt:9 (find_package)


CMake Error at /usr/lib/x86_64-linux-gnu/cmake/eCAL/eCALConfig.cmake:71 (include):
  include could not find load file:

    /usr/local/lib/x86_64-linux-gnu/cmake/eCAL/helper_functions/ecal_install_functions.cmake
Call Stack (most recent call first):
  CMakeLists.txt:9 (find_package)


CMake Error at /usr/lib/x86_64-linux-gnu/cmake/eCAL/eCALConfig.cmake:72 (include):
  include could not find load file:

    /usr/local/lib/x86_64-linux-gnu/cmake/eCAL/eCALTargets.cmake
Call Stack (most recent call first):
  CMakeLists.txt:9 (find_package)

```

**Comment by rex-schilasky** [comment]
Hi @prakhar114. Did you checkout the submodules like described [here](https://eclipse-ecal.github.io/ecal/development/building_ecal_from_source.html) ?

```
git clone https://github.com/eclipse-ecal/ecal.git
cd ecal
git submodule init
git submodule update
```

**Reply by prakhar114** [reply]
Getting this while make -j4

```
/usr/bin/ld: /usr/local/lib/libprotobuf.a(arena.cc.o): relocation R_X86_64_TPOFF32 against hidden symbol `_ZN6google8protobuf8internal15ThreadSafeArena13thread_cache_E' can not be used when making a shared object
collect2: error: ld returned 1 exit status
make[2]: *** [ecal/core/CMakeFiles/core.dir/build.make:960: lib/libecal_core.so.5.11.0] Error 1
make[1]: *** [CMakeFiles/Makefile2:3131: ecal/core/CMakeFiles/core.dir/all] Error 2
make: *** [Makefile:152: all] Error 2
```

**Reply by FlorianReimold** [reply]
Hi prakhar,

Please elaborate:
- What you are trying to compile
- How you compile it (e.g. your CMake options)
- What OS you are on
- What eCAL Version you are using
- Where you got eCAL from

The error you are getting is because you are linking a static version of libprotobuf into a shared object, while libprotobuf.a probably hasn't been compile with -fPIC .

**Reply by prakhar114** [reply]
I've tried building ecal from source described over [here](https://eclipse-ecal.github.io/ecal/development/building_ecal_from_source.html). 
I did the following
```
git clone https://github.com/eclipse-ecal/ecal.git
cd ecal
git submodule init
git submodule update
```

I'm on Ubuntu 20.04
```
sudo apt-get install git cmake doxygen graphviz build-essential zlib1g-dev qtbase5-dev libhdf5-dev libprotobuf-dev libprotoc-dev protobuf-compiler libcurl4-openssl-dev libqwt-qt5-dev libyaml-cpp-dev
```
```
mkdir _build
cd _build
cmake .. -DCMAKE_BUILD_TYPE=Release -DECAL_THIRDPARTY_BUILD_PROTOBUF=OFF -DECAL_THIRDPARTY_BUILD_CURL=OFF -DECAL_THIRDPARTY_BUILD_HDF5=OFF -DECAL_THIRDPARTY_BUILD_QWT=OFF
make -j4
```
After `make -j4` I'm getting the error below.
```
/usr/bin/ld: /usr/local/lib/libprotobuf.a(arena.cc.o): relocation R_X86_64_TPOFF32 against hidden symbol `_ZN6google8protobuf8internal15ThreadSafeArena13thread_cache_E' can not be used when making a shared object
collect2: error: ld returned 1 exit status
make[2]: *** [ecal/core/CMakeFiles/core.dir/build.make:960: lib/libecal_core.so.5.11.0] Error 1
make[1]: *** [CMakeFiles/Makefile2:3131: ecal/core/CMakeFiles/core.dir/all] Error 2
make: *** [Makefile:152: all] Error 2
```

**Reply by FlorianReimold** [reply]
OK, so for some unknown reason your Ubuntu 20.04 finds a static version of libprotobuf. Even though eCAL can be compiled statically against libprotobuf (we even do that as default on Windows), the `libprotobuf-dev` system package is compiled with the wrong flags and therefore fails to link into a shared object like `libecal_core.so`.

You can try to fix the issue, by adding `-DProtobuf_USE_STATIC_LIBS=OFF` to your cmake command line. Maybe also delete your build folder before re-cmaking. While CMake is configuring the project, it will at some point tell you about the Protobuf libraries that it found. At least on a vanilla Ubuntu 20.04 it will show:

```bash
[...]
-- Found Protobuf: /usr/lib/x86_64-linux-gnu/libprotobuf.so;-pthread (found version "3.6.1") 
[...]
```

Keep an eye on that line. It should show a path to an `.so` file with version 3.6.1. If you are getting anything else you may have a non-standard protobuf version installed, that maybe doesn't ship with a shared object?
